compiler/damlc/daml-stdlib-src/DA/Upgrade.daml

```diff
@@ -10,41 +10,88 @@ daml 1.2
 module DA.Upgrade
   ( conv
   , iso
+  , Convertable(..)
+  , GenConvertable
+  , Upgrade(..)
+  , Rollback(..)
   )
   where
 
 import DA.Generics
-
--- | Convert data types that are isomorphic and have the same meta-data up to package id.
-conv : (Generic a repA, Generic b repB, Conv repA repB) => a -> b
+import DA.Next.Set
+
+----------------------------------------------------------------------------------------------------
+-- Upgrade templates
+----------------------------------------------------------------------------------------------------
+
+template (Template i, Template o, Convertable i o) => Upgrade i o
+    with
+        op : Party
+    where
+        signatory op
+        nonconsuming choice DoUpgrade: ContractId o
+            with
+                inC : ContractId i
+                sigs : [Party]
+            controller sigs
+                do
+                    d <- fetch inC
+                    assert $ fromList sigs == fromList (signatory d)
+                    create $ convert d
+
+template (Template i, Template o, Convertable o i) => Rollback i o
+    with
+        op : Party
+    where
+        signatory op
+        nonconsuming choice DoRollback: ContractId i
+            with
+                inC : ContractId o
+                sigs : [Party]
+            controller sigs
+                do
+                    d <- fetch inC
+                    assert $ fromList sigs == fromList (signatory d)
+                    create $ convert d
+
+----------------------------------------------------------------------------------------------------
+-- Data type conversion
+----------------------------------------------------------------------------------------------------
+
+-- | Types, for which there exists a conversion.
+class Convertable a b where
+  convert : a -> b
```

Comment:
Can we add a default implementation as in
https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#default-method-signatures
for this? This would have the advantage, that we can just dump a
```haskell
instance Convertible Old.X New.X
```
into the auto generated code and it passes if the conversion via generic works and fails with a "method missing" error otherwise.
If that works, we can remove the `conv` function. If not, we should probably rename it to `genConvert`.

Reply:
that would be cool, I'll give it a try